### DISCUSSION TITLE
Optimistic updates not updating cache as expected

### DISCUSSION THREAD

**Discussion by 01jonathanf** [urql-graphql/urql#1645] [Q&A] [answered]
Hi,

I'm applying the following optimistic configuration in the cacheexchange then re-executing my query on completion (with cache-only setting) but it doesn't seem to be updating the cache at all.

```

optimistic: {
        updateSyllable: (variables, cache, info) => ({
          __typename: "Syllable",
          id: variables.input["filter"].id,
          active_mnemonic: variables.input["set"].active_mnemonic
        })
      } 

```

I've also tried adding a writeFragment config to the updates like so but it makes no difference:

```
updates: {
        Mutation: {
          updateSyllable(_result, args, cache, _info) {
            const fragment = gql`
              fragment _ on Syllable {
                __typename
                id

                active_mnemonic {
                  id
                }
              }
           ` ;

            cache.writeFragment(fragment, {
              __typename: "Syllable",
              id: args.input["filter"].id,
              active_mnemonic: {
                id: args.input["set"].active_mnemonic.id
              }
            });
          }
        }
      },

```

I'm wondering what I'm missing here to be able to make this work?

**Comment by kitten** [comment, marked as answer]
Most of the time this is because your optimistic configuration there is missing some data and a subsequent query then experiences a cache miss that can only be resolved once the mutation completes. Internally it'll basically try to fetch data from the API, but will then be forced to wait until the mutation completes. We currently don't have a warning for this because it's a little tricky to detect automatically.

But if you look at the urql devtools you should actually see the cache miss come up.

To me it's actually looking like `active_mnemonic` is missing a `__typename` property. That being said, it can have various causes depending on what fields your queries will need to have available.

You're also right in that writing an updater shouldn't be needed. In fact, in this case your updater won't do anything additionally to your optimistic function.
